File: 2610-convert-an-array-into-a-2d-array-with-conditions/2610-convert-an-array-into-a-2d-array-with-conditions.cpp
class Solution {
public:
    vector<vector<int>> findMatrix(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int, int> map;
        vector<vector<int>> result;
        for(int i=0; i<n; i++){
            int freq = map[nums[i]];
            if(freq == result.size()){
                result.push_back({});
            }
            result[freq].push_back(nums[i]);
            map[nums[i]]++;
        }
        return result;
    }
};

File: 2610-convert-an-array-into-a-2d-array-with-conditions/2610-convert-an-array-into-a-2d-array-with-conditions.java
class Solution {
    public List<List<Integer>> findMatrix(int[] nums) {
        int n = nums.length;
        Map<Integer, Integer> map = new HashMap<>();
        List<List<Integer>> result = new ArrayList<>();
        for(int i=0; i<n; i++){
            int freq = map.getOrDefault(nums[i], 0);
            if(freq == result.size()){
                result.add(new ArrayList<>());
            }
            result.get(freq).add(nums[i]);
            map.put(nums[i], freq+1);
        }
        return result;
    }
}